src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,361 @@
+                children = new List<AnalyzedNode>();
+                GetChildren(children);
+                return RecursivePattern(
+                    children.OfType<TypePattern>().SingleOrDefault()?.Type,
+                    null,
+                    PropertyPatternClause(SeparatedList(children.OfType<PatternMatch>().Select(match => match.AsSubpatternSyntax()))),
+                    children.OfType<VarPattern>().SingleOrDefault()?.AsVariableDesignationSyntax());
+            }
+
+            public override ExpressionSyntax AsExpressionSyntax()
+            {
+                return BinaryExpression(
+                    SyntaxKind.LogicalAndExpression,
+                    Left.AsExpressionSyntax(),
+                    Right.AsExpressionSyntax());
+            }
+
+            public CasePatternSwitchLabelSyntax AsCasePatternSwitchLabelSyntax()
+            {
+                 return CasePatternSwitchLabel(AsPatternSyntax(out var children),
+                    children.OfType<Evaluation>().SingleOrDefault()?.AsWhenClauseSyntax(),
+                    Token(SyntaxKind.ColonToken));
+            }
+
+            private static AnalyzedNode UnionCore(Conjuction conjuction, PatternMatch match)
+            {
+                if (conjuction.Left.Contains(match.Expression))
+                {
+                    return new Conjuction(Union(conjuction.Left, match), conjuction.Right);
+                }
+
+                if (conjuction.Right.Contains(match.Expression))
+                {
+                    return new Conjuction(conjuction.Left, Union(conjuction.Right, match));
+                }
+
+                return new Conjuction(conjuction, match);
+            }
+
+            private static AnalyzedNode UnionCore(PatternMatch leftMatch, PatternMatch rightMatch)
+            {
+                if (AreEquivalent(leftMatch.Expression, rightMatch.Expression))
+                {
+                    return new PatternMatch(leftMatch.Expression, Union(leftMatch.Pattern, rightMatch.Pattern));
+                }
+
+                if (leftMatch.Pattern.Contains(rightMatch.Expression))
+                {
+                    return new PatternMatch(leftMatch.Expression, Union(leftMatch.Pattern, rightMatch));
+                }
+
+                if (rightMatch.Pattern.Contains(leftMatch.Expression))
+                {
+                    return new PatternMatch(rightMatch.Expression, Union(rightMatch.Pattern, leftMatch));
+                }
+
+                return new Conjuction(leftMatch, rightMatch);
+            }
+
+            private static AnalyzedNode UnionCore(VarPattern var, PatternMatch match)
+            {
+                if (var.Contains(match.Expression))
+                {
+                    return new Conjuction(var, match.Pattern);
+                }
+
+                return new Conjuction(var, match);
+            }
+
+            private static AnalyzedNode Union(AnalyzedNode left, AnalyzedNode right)
```

Comment:
Note to self: this should be really `Intersect` (similar to TS intersection types which is denoted by `T & U`). Later when we have `or` patterns, `||` would produce a `Disjunction` node which will be reduced by taking the Union of sides.